web/pdf_find_controller.js

```diff
@@ -47,47 +47,177 @@ const CHARACTERS_TO_NORMALIZE = {
+  // shift.
+  // Let say that the last entry in the array is [x, s] and we have a
+  // substitution at index y (old string) which will replace o chars by n chars.
+  // Firstly, if o === n, then no need to add a new entry: the shift is
+  // the same.
+  // Secondly, if o < n, then we push the n - o elements:
+  // [y - (s - 1), s - 1], [y - (s - 2), s - 2], ...
+  // Thirdly, if o > n, then we push the element: [y - (s - n), o + s - n]
+
+  // Collect diacritics length and positions.
+  const rawDiacriticsPositions = [];
+  let m;
+  while ((m = DIACRITICS_REG_EXP.exec(text)) !== null) {
+    rawDiacriticsPositions.push([m[0].length, m.index]);
+  }
+
+  let normalized = text.normalize("NFD");
+  const positions = [[0, 0]];
+  let k = 0;
+  let shift = 0;
+  let shiftOrigin = 0;
+  let eol = 0;
+  let hasDiacritics = false;
+
+  normalized = normalized.replace(
+    normalizationRegex,
+    (match, p1, p2, p3, p4, i) => {
+      i -= shiftOrigin;
+      if (p1) {
+        // Maybe fractions or quotations mark...
+        const replacement = CHARACTERS_TO_NORMALIZE[match];
+        const jj = replacement.length;
+        for (let j = 1; j < jj; j++) {
+          positions.push([i - shift + j, shift - j]);
+        }
+        shift -= jj - 1;
+        return replacement;
+      }
+
+      if (p2) {
+        // "X-\n" is removed because an hypen at the end of a line
+        // with not a space before is likely here to mark a break
+        // in a word.
+        positions.push([i - shift, 1 + shift]);
+        shift += 1;
+        shiftOrigin += 1;
+        eol += 1;
+        return p2.charAt(0);
+      }
+
+      if (p3) {
+        // eol is replaced by space: "foo\nbar" is likely equivalent to
+        // "foo bar".
+        positions.push([i - shift + 1, shift - 1]);
+        shift -= 1;
+        shiftOrigin += 1;
+        eol += 1;
+        return " ";
+      }
+
+      // Diacritics.
+      hasDiacritics = true;
+      let jj = p4.length;
+      if (i + eol === rawDiacriticsPositions?.[k]?.[1]) {
+        jj -= rawDiacriticsPositions[k][0];
+        ++k;
+      }
+
+      for (let j = 1; j < jj + 1; j++) {
+        // i is the position of the first diacritic
+        // so (i - 1) is the position for the letter before.
+        positions.push([i - 1 - shift + j, shift - j]);
+      }
+      shift -= jj;
+      shiftOrigin += jj;
+
+      return p4;
     }
-    return normalizedCh;
-  });
+  );
 
-  return [normalizedText, diffs];
+  positions.push([normalized.length, shift]);
+
+  return [normalized, positions, hasDiacritics];
 }
 
-// Determine the original, non-normalized, match index such that highlighting of
-// search results is correct in the `textLayer` for strings containing e.g. "½"
-// characters; essentially "inverting" the result of the `normalize` function.
-function getOriginalIndex(matchIndex, diffs = null) {
-  if (!diffs) {
-    return matchIndex;
+function getOriginalIndex(positions, pos, len) {
```

Comment:
Any reason for renaming `diffs` to `positions` here?

Also, please note that the old function handled the `diffs` being `null` which can still happen for pages where text-extraction fails. Hence you'll most likely need to *still* handle that case here, to prevent any errors being thrown and breaking searching; please see the old function.